src/treelearner/monotone_constraints.hpp

```diff
@@ -16,52 +16,339 @@
+
+  BasicConstraintEntry *clone() const final {
+    return new BasicConstraintEntry(*this);
+  };
+
+  void Reset() final {
     min = -std::numeric_limits<double>::max();
     max = std::numeric_limits<double>::max();
   }
 
-  void UpdateMin(double new_min) { min = std::max(new_min, min); }
+  void UpdateMin(double new_min) final { min = std::max(new_min, min); }
 
-  void UpdateMax(double new_max) { max = std::min(new_max, max); }
+  void UpdateMax(double new_max) final { max = std::min(new_max, max); }
 
-  bool UpdateMinAndReturnBoolIfChanged(double new_min) {
+  bool UpdateMinAndReturnBoolIfChanged(double new_min) final {
     if (new_min > min) {
       min = new_min;
       return true;
     }
     return false;
   }
 
-  bool UpdateMaxAndReturnBoolIfChanged(double new_max) {
+  bool UpdateMaxAndReturnBoolIfChanged(double new_max) final {
     if (new_max < max) {
       max = new_max;
       return true;
     }
     return false;
   }
+
+  BasicConstraint LeftToBasicConstraint() const final { return *this; }
+
+  BasicConstraint RightToBasicConstraint() const final { return *this; }
+
+  FeatureConstraint *GetFeatureConstraint(int) final { return this; }
+};
+
+struct FeatureMinOrMaxConstraints {
+  std::vector<double> constraints;
+  // the constraint number i is valid on the slice
+  // [thresholds[i]:threshold[i+1])
+  // if threshold[i+1] does not exist, then it is valid for thresholds following
+  // threshold[i]
+  std::vector<uint32_t> thresholds;
+
+  FeatureMinOrMaxConstraints() {
+    constraints.reserve(32);
+    thresholds.reserve(32);
+  }
+
+  size_t Size() const { return thresholds.size(); }
+
+  explicit FeatureMinOrMaxConstraints(double extremum) {
+    constraints.reserve(32);
+    thresholds.reserve(32);
+
+    constraints.push_back(extremum);
+    thresholds.push_back(0);
+  }
+
+  void Reset(double extremum) {
+    constraints.resize(1);
+    constraints[0] = extremum;
+    thresholds.resize(1);
+    thresholds[0] = 0;
+  }
+
+  void UpdateMin(double min) {
+    for (unsigned int j = 0; j < constraints.size(); j++) {
+      if (min > constraints[j]) {
+        constraints[j] = min;
+      }
+    }
+  }
+
+  void UpdateMax(double max) {
+    for (unsigned int j = 0; j < constraints.size(); j++) {
+      if (max < constraints[j]) {
+        constraints[j] = max;
+      }
+    }
+  }
+};
+
+struct CumulativeFeatureConstraint {
+  std::vector<uint32_t> thresholds_min_constraints;
+  std::vector<uint32_t> thresholds_max_constraints;
+  std::vector<double> cumulative_min_constraints_left_to_right;
+  std::vector<double> cumulative_min_constraints_right_to_left;
+  std::vector<double> cumulative_max_constraints_left_to_right;
+  std::vector<double> cumulative_max_constraints_right_to_left;
+  unsigned int index_min_constraints_left_to_right;
+  unsigned int index_min_constraints_right_to_left;
+  unsigned int index_max_constraints_left_to_right;
+  unsigned int index_max_constraints_right_to_left;
```

Comment:
I'm not cpp programmer, but seems that changing `unsigned int` to `size_t` should fix the following compiler warnings:
```
d:\a\1\s\src\treelearner\monotone_constraints.hpp(198): warning C4267: '=': conversion from 'size_t' to 'unsigned int', possible loss of data (compiling source file D:\a\1\s\src\treelearner\data_parallel_tree_learner.cpp) [D:\a\1\s\build\_lightgbm.vcxproj]
d:\a\1\s\src\treelearner\monotone_constraints.hpp(200): warning C4267: '=': conversion from 'size_t' to 'unsigned int', possible loss of data (compiling source file D:\a\1\s\src\treelearner\data_parallel_tree_learner.cpp) [D:\a\1\s\build\_lightgbm.vcxproj]
d:\a\1\s\src\treelearner\monotone_constraints.hpp(202): warning C4267: '=': conversion from 'size_t' to 'unsigned int', possible loss of data (compiling source file D:\a\1\s\src\treelearner\data_parallel_tree_learner.cpp) [D:\a\1\s\build\_lightgbm.vcxproj]
d:\a\1\s\src\treelearner\monotone_constraints.hpp(204): warning C4267: '=': conversion from 'size_t' to 'unsigned int', possible loss of data (compiling source file D:\a\1\s\src\treelearner\data_parallel_tree_learner.cpp) [D:\a\1\s\build\_lightgbm.vcxproj]
```